src/server.c

```diff
@@ -192,6 +203,18 @@ static bool handle_server_startup(PgSocket *server, PktHdr *pkt)
 	return res;
 }
 
+/*
+ * server_pool_mode returns the pool_mode for the server. It specifically
+ * forces session pooling if the server is a replication connection, because
+ * replication connections require session pooling to work correctly.
+ */
+int server_pool_mode(PgSocket *server)
+{
+	if (server->replication)
+		return POOL_SESSION;
+	return pool_pool_mode(server->pool);
+}
+
```

Comment:
I'm worried that if we have both `server_pool_mode()` and still the old `pool_pool_mode()`, it will be difficult to enforce that the right one is called in each situation.

I wonder if we could alleviate that by allocating a `PgPool` structure for each replication connection, but only allow one connection through it. Then we can keep using `pool_pool_mode()`, and maybe also have fewer changes elsewhere.  (For example, this patch doesn't appear to have anything for statistics or `SHOW` commands. I wonder if this still works sensibly for connections that are not associated with pools. And of course, we do want some sort of monitoring support in any case.)

Reply:
It's a reasonable worry, because I could definitely see that be an issue. To avoid the issue I added a warning to the function comment of `pool_pool_mode()` and renamed it to `probably_wrong_pool_pool_mode()`, because it should basically never be used directly (the only place where it should be used is in `SHOW pools`).

I did that instead of your proposal, because I do want to keep the current general design. The reason for that is that this is more in line with the future plan I have of allowing a client to choose if they want transaction or session pooling: https://github.com/pgbouncer/pgbouncer/issues/997 Basically for that near-future plan the pooling mode would be a property of the connection, not of the pool. Just like it is a property of the connection now for replication connections. 

Regarding statistics, I think it would make sense to add a replication and pool_mode column to `show servers/clients`. But the existing reporting also works for replication connections.

Reply:
FYI I added a replication column to `SHOW CLIENTS/SERVERS`.  I think that should be enough info for debugging any issues.